src/fields/surface_depth.cc

```diff
@@ -0,0 +1,126 @@
+/*!
+ *
+﻿ * Copyright (C) 2015 Technical University of Liberec.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 3 as published by the
+ * Free Software Foundation. (http://www.gnu.org/licenses/gpl-3.0.en.html)
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ *
+ *
+ * @file    surface_depth.cc
+ * @brief
+ */
+
+#include <vector>
+#include <armadillo>
+#include <limits>
+
+#include "fields/surface_depth.hh"
+#include "mesh/mesh.h"
+
+
+SurfaceDepth::SurfaceDepth(const Mesh *mesh, std::string surface_region, std::string surface_direction)
+{
+    this->create_projection_matrix( arma::vec3(surface_direction) );
+    this->construct_bih_tree( const_cast<Mesh*>(mesh), surface_region);
+}
+
+
+void SurfaceDepth::create_projection_matrix(arma::vec3 surface_vec)
+{
+    if (surface_vec(2)==0) {
+        THROW( ExcSurfaceProjection() << EI_Message("Vertical plane of surface is forbidden!") );
+    }
+    surface_norm_vec_ = surface_vec / arma::norm(surface_vec, 2); // normalize to size == 1
+
+    arma::vec3 ex("1 0 0"), ey("0 1 0");
+    arma::vec3 ta = ex - arma::dot(ex, surface_norm_vec_) * surface_norm_vec_;
+    ta /= arma::norm(ta, 2); // normalize
+    arma::vec3 tb = ey - arma::dot(ey, surface_norm_vec_) * surface_norm_vec_;
+    tb /= arma::norm(tb, 2); // normalize
+    m_ = arma::mat(2,3);
+    m_.row(0) = ta.t();
+    m_.row(1) = tb.t();
+}
+
+
+void SurfaceDepth::construct_bih_tree(Mesh *mesh, std::string surface_region)
+{
+	std::vector<BoundingBox> boxes;
+	nodes_coords_.clear();
+
+	RegionSet region_set = mesh->region_db().get_region_set(surface_region);
+    if (region_set.size() == 0)
+        THROW( RegionDB::ExcUnknownSet() << RegionDB::EI_Label(surface_region) );
+
+    // make element boxes
+    unsigned int i=0;
+    unsigned int i_node;
+    arma::vec3 project_node("0 0 0");
+    for( ElementFullIter it = mesh->bc_elements.begin(); it != mesh->bc_elements.end(); ++it) {
```

Comment:
We should iterate through all elements to allow projections to internal interfaces. Depends on
changes in mesh. We should merge mesh changes ASAP.